File: obj2ms3dascii.lua
-- Ref : Need some help with Milkshape 3D ascii format - http://www.gamedev.net/topic/464065-need-some-help-with-milkshape-3d-ascii-format/

----------------

-- Ref : Luaで文字列の分割を行なう(split) - http://symfoware.blog68.fc2.com/blog-entry-455.html
function split( _str, _sep )
	if string.find( _str, _sep ) == nil then
		return { _str }
	end

	local ret = {}
	local pat = "(.-)" .. _sep .. "()"
	local lastPos
	for part, pos in string.gmatch( _str, pat ) do
		table.insert( ret, part )
		lastPos = pos
	end
	table.insert( ret, string.sub( _str, lastPos ) )
	return ret
end

-----------------

-- load .obj
objPath = arg[1]
if objPath == nil then
	print( "This script needs .obj file!\nUsage : > obj2ms3dascii.lua cube.obj\n" )
	os.exit()
else
	objFile = io.open( objPath, "r" )
	if objFile == nil then
		print( "not found : " .. objPath )
		os.exit()
	end
end

vertices = {}
vertices["v"] = {}
vertices["vt"] = {}
vertices["vn"] = {}
objects = {}
currentObject = 0
mtlPath = nil

-- read .obj
for line in objFile:lines() do
	li = split( line, " " )
	if li[1] == "mtllib" then
		mtlPath = li[2]
	elseif li[1] == "o" then
		currentObject = currentObject + 1
		objects[currentObject] = {}
		objects[currentObject]["name"] = li[2]
		objects[currentObject]["f"] = {}
	elseif li[1] == "v" then table.insert( vertices["v"], { li[2], li[3], li[4] } )
	elseif li[1] == "vt" then table.insert( vertices["vt"], { li[2], li[3] } )
	elseif li[1] == "vn" then table.insert( vertices["vn"], { li[2], li[3], li[4] } )
	elseif li[1] == "f" then
		for iLi = 4, #li do
			table.insert( objects[currentObject]["f"], { li[2], li[iLi-1], li[iLi] } )
		end
	elseif li[1] == "usemtl" then objects[currentObject]["usemtl"] = li[2]
	end
end
objFile:close()

--------------------------

-- load .mtl
if mtlPath == nil then
	print( "mtllib not found" )
	os.exit()
else
	mtlFile = io.open( mtlPath, "r" )
	if mtlFile == nil then
		print( "not found : " .. mtlPath )
		os.exit()
	end
end

materials = {}
currentMaterial = 0

-- read .mtl
for line in mtlFile:lines() do
	li = split( line, " " )
	if li[1] == "newmtl" then
		currentMaterial = currentMaterial + 1
		materials[currentMaterial] = {}
		materials[currentMaterial]["name"] = li[2]
		materials[currentMaterial]["Ka"] = { "0.000", "0.000", "0.000" }
		materials[currentMaterial]["Kd"] = { "0.000", "0.000", "0.000" }
		materials[currentMaterial]["Ks"] = { "0.000", "0.000", "0.000" }
		materials[currentMaterial]["Ke"] = { "0.000", "0.000", "0.000" }
		materials[currentMaterial]["Ns"] = "0.000"
		materials[currentMaterial]["d"] = "0.000"
		materials[currentMaterial]["map_Kd"] = ""
		materials[currentMaterial]["map_d"] = ""
	elseif li[1] == "Ka" then materials[currentMaterial]["Ka"] = { li[2], li[3], li[4] }
	elseif li[1] == "Kd" then materials[currentMaterial]["Kd"] = { li[2], li[3], li[4] }
	elseif li[1] == "Ks" then materials[currentMaterial]["Ks"] = { li[2], li[3], li[4] }
	elseif li[1] == "Ke" then materials[currentMaterial]["Ke"] = { li[2], li[3], li[4] }
	elseif li[1] == "Ns" then materials[currentMaterial]["Ns"] = li[2]
	elseif li[1] == "d" then materials[currentMaterial]["d"] = li[2]
	elseif li[1] == "map_Kd" then materials[currentMaterial]["map_Kd"] = li[2]
	elseif li[1] == "map_d" then materials[currentMaterial]["map_d"] = li[2]
	end
end
mtlFile:close()

-------------------

-- check obj index usage
indexUsage = {}
indexUsage[1] = not not tonumber( split( objects[1]["f"][1][1], "/" )[1] ) -- v
indexUsage[2] = not not tonumber( split( objects[1]["f"][1][1], "/" )[2] ) -- vt
indexUsage[3] = not not tonumber( split( objects[1]["f"][1][1], "/" )[3] ) -- vn

-------------------

-- obj v/vt to ms3dascii vertex
vertices["ms"] = {}
vertexIndex = {}

-- attribute vertices with textureCoods
for iObject, vObject in ipairs( objects ) do
	for iFace, vFace in ipairs( vObject["f"] ) do
		for iIndex, vIndex in ipairs( vFace ) do
			if not vertexIndex[vIndex] then
				ind = split( vIndex, "/" )
				if indexUsage[2] then
					table.insert( vertices["ms"], { vertices["v"][tonumber(ind[1])][1], vertices["v"][tonumber(ind[1])][2], vertices["v"][tonumber(ind[1])][3], vertices["vt"][tonumber(ind[2])][1], vertices["vt"][tonumber(ind[2])][2] } ) -- x, y, z, u, v
				else
					table.insert( vertices["ms"], { vertices["v"][tonumber(ind[1])][1], vertices["v"][tonumber(ind[1])][2], vertices["v"][tonumber(ind[1])][3], 0.0, 0.0 } ) -- x, y, z, u, v
				end
				vertexIndex[vIndex] = #vertices["ms"]
			end
		end
	end
end

-- obj vn to ms3dascii normal
vertices["msn"] = {}

if indexUsage[3] then
	for iNormal, vNormal in ipairs( vertices["vn"] ) do
		table.insert( vertices["msn"], { vNormal[1], vNormal[2], vNormal[3] } )
	end
else
	for iObject, vObject in ipairs( objects ) do
		vertices["msn"][iObject] = {}
		for iFace, vFace in ipairs( vObject["f"] ) do
			-- cross product, compute normal
			vf = {}
			vf[1] = split( vFace[1], "/" )[1]
			vf[2] = split( vFace[2], "/" )[1]
			vf[3] = split( vFace[3], "/" )[1]
			a = { vertices["v"][tonumber(vf[2])][1]-vertices["v"][tonumber(vf[1])][1], vertices["v"][tonumber(vf[2])][2]-vertices["v"][tonumber(vf[1])][2], vertices["v"][tonumber(vf[2])][3]-vertices["v"][tonumber(vf[1])][3] }
			b = { vertices["v"][tonumber(vf[3])][1]-vertices["v"][tonumber(vf[2])][1], vertices["v"][tonumber(vf[3])][2]-vertices["v"][tonumber(vf[2])][2], vertices["v"][tonumber(vf[3])][3]-vertices["v"][tonumber(vf[2])][3] }
			n = { a[2]*b[3] - a[3]*b[2], a[3]*b[1] - a[1]*b[3], a[1]*b[2] - a[2]*b[1] }
			nLength = -math.sqrt( n[1]*n[1] + n[2]*n[2] + n[3]*n[3] )
			n = { n[1]/nLength, n[2]/nLength, n[3]/nLength }
			table.insert( vertices["msn"][iObject], n )
		end
	end
end

-------------------

-- write ms3dascii
out = [[// MilkShape 3D ASCII
// Converted from .obj with obj2ms3dascii.lua by FMS_Cat
// https://github.com/FMS-Cat/obj2ms3dascii
Frames: 30
Frame: 1
]]

out = out .. string.format( "Meshes: %d\n", #objects )
for iObject, vObject in ipairs( objects ) do

	for iMaterial, vMaterial in ipairs( materials ) do
		if vObject["usemtl"] == vMaterial["name"] then
			materialIndex = iMaterial-1
			break
		end
	end
	out = out .. string.format( "\"%s\" 0 %d\n", vObject["name"], materialIndex )

	out = out .. string.format( "%d\n", #vertices["ms"] )
	for iVertex, vVertex in ipairs( vertices["ms"] ) do
		out = out .. string.format( "0 %f %f %f %f %f -1\n", vVertex[1], vVertex[2], vVertex[3], vVertex[4], vVertex[5] );
	end

	if indexUsage[3] then
		out = out .. string.format( "%d\n", #vertices["msn"] )
		for iNormal, vNormal in ipairs( vertices["msn"] ) do
			out = out .. string.format( "%f %f %f\n", vNormal[1], vNormal[2], vNormal[3] );
		end
	else
		out = out .. string.format( "%d\n", #vertices["msn"][iObject] )
		for iNormal, vNormal in ipairs( vertices["msn"][iObject] ) do
			out = out .. string.format( "%f %f %f\n", vNormal[1], vNormal[2], vNormal[3] )
		end
	end

	out = out .. string.format( "%d\n", #vObject["f"] )
	if indexUsage[3] then
		for iFace, vFace in ipairs( vObject["f"] ) do
			ind = {}
			ind[1] = split( vFace[1], "/" )
			ind[2] = split( vFace[2], "/" )
			ind[3] = split( vFace[3], "/" )
			out = out .. string.format( "0 %d %d %d %d %d %d 1\n", vertexIndex[vFace[1]]-1, vertexIndex[vFace[2]]-1, vertexIndex[vFace[3]]-1, ind[1][3]-1, ind[2][3]-1, ind[3][3]-1 );
		end
	else
		for iFace, vFace in ipairs( vObject["f"] ) do
			ind = {}
			ind[1] = split( vFace[1], "/" )
			ind[2] = split( vFace[2], "/" )
			ind[3] = split( vFace[3], "/" )
			out = out .. string.format( "0 %d %d %d %d %d %d 1\n", vertexIndex[vFace[1]]-1, vertexIndex[vFace[2]]-1, vertexIndex[vFace[3]]-1, iFace-1, iFace-1, iFace-1 );
		end
	end
end

out = out .. string.format( "Materials: %d\n", #materials )
for iMaterial, vMaterial in ipairs( materials ) do
	out = out .. string.format( "\"%s\"\n", vMaterial["name"] )
	out = out .. string.format( "%f %f %f 1.000000\n", vMaterial["Ka"][1], vMaterial["Ka"][2], vMaterial["Ka"][3] )
	out = out .. string.format( "%f %f %f 1.000000\n", vMaterial["Kd"][1], vMaterial["Kd"][2], vMaterial["Kd"][3] )
	out = out .. string.format( "%f %f %f 1.000000\n", vMaterial["Ks"][1], vMaterial["Ks"][2], vMaterial["Ks"][3] )
	out = out .. string.format( "%f %f %f 1.000000\n", vMaterial["Ke"][1], vMaterial["Ke"][2], vMaterial["Ke"][3] )
	out = out .. string.format( "%f\n", vMaterial["Ns"] )
	out = out .. string.format( "%f\n", vMaterial["d"] )
	out = out .. string.format( "\"%s\"\n", vMaterial["map_Kd"] )
	out = out .. string.format( "\"%s\"\n", vMaterial["map_d"] )
end

out = out .. [[Bones: 1
"Bone01"
""
0 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000
1
1.000000 0.000000 0.000000 0.000000
1
1.000000 0.000000 0.000000 0.000000

GroupComments: 0
MaterialComments: 0
BoneComments: 0
ModelComment: 0]]
outFile = io.open( objPath .. ".txt", "w" )
outFile:write( out )
outFile:close();

File: obj2ms3dascii.mjs
/**
 * Import an obj model and export as a ms3dascii model.
 * Ref : Need some help with Milkshape 3D ascii format - http://www.gamedev.net/topic/464065-need-some-help-with-milkshape-3d-ascii-format/
 * @param {string} obj The entire file of the obj
 * @param {string} mtl The entire file of the mtl
 */
export function obj2ms3dascii( obj, mtl ) {
  // -- read .obj ----------------------------------------------------------------------------------
  const vertices = {
    v: [],
    vt: [],
    vn: [],
  };

  let currentObject = null;
  const unfilteredObjects = []; // all found objects before being filtered to remove "invalid" objects

  // let mtlPath = null;

  if ( obj == null ) {
    throw new Error( 'obj2ms3dascii: No obj file detected' );
  }

  obj.split( '\n' ).forEach( ( line ) => {
    const li = line.split( ' ' );

    if ( li[ 0 ] === 'mtllib' ) {
      // mtlPath = li[ 1 ];
    } else if ( li[ 0 ] === 'o' || li[ 0 ] === 'g' ) {
      currentObject = {
        name: li[ 1 ],
        f: [],
      };
      unfilteredObjects.push( currentObject );
    } else if ( li[ 0 ] === 'v' ) {
      const vertex = [ li[ 1 ], li[ 2 ], li[ 3 ] ].map( ( v ) => parseFloat( v ) );
      vertices.v.push( vertex );
    } else if ( li[ 0 ] === 'vt' ) {
      const vertexUv = [ li[ 1 ], li[ 2 ] ].map( ( v ) => parseFloat( v ) );
      vertices.vt.push( vertexUv );
    } else if ( li[ 0 ] === 'vn' ) {
      const vertexNormal = [ li[ 1 ], li[ 2 ], li[ 3 ] ].map( ( v ) => parseFloat( v ) );
      vertices.vn.push( vertexNormal );
    } else if ( li[ 0 ] === 'f' ) {
      // `f 1//1 2//1 4//1 3//1`
      for ( let iLi = 3; iLi < li.length; iLi ++ ) {
        const face = [ li[ 1 ], li[ iLi - 1 ], li[ iLi ] ].map(
          ( i ) => i.split( '/' ).map( ( i ) => i !== '' ? parseInt( i ) : null )
        );
        currentObject.f.push( face );
      }
    } else if ( li[ 0 ] === 'usemtl' ) {
      currentObject.usemtl = li[ 1 ];
    }
  } );
  
  // Filter out "invalid" objects
  const objects = unfilteredObjects.filter( ( x ) => x.f.length > 0 );

  // -- read .mtl ----------------------------------------------------------------------------------
  // if ( mtlPath == null ) {
  //   throw new Error( 'obj2ms3dascii: mtllib not found' );
  // }

  let currentMaterial = {};
  const materials = [];

  mtl.split( '\n' ).forEach( ( line ) => {
    const li = line.split( ' ' );

    if ( li[ 0 ] === 'newmtl' ) {
      currentMaterial = {
        name: li[ 1 ],
        Ka: [ 0, 0, 0 ],
        Kd: [ 0, 0, 0 ],
        Ks: [ 0, 0, 0 ],
        Ke: [ 0, 0, 0 ],
        Ns: 0,
        d: 0,
        map_Kd: '',
        map_d: '',
      };
      materials.push( currentMaterial );
    } else if ( li[ 0 ] === 'Ka' ) {
      currentMaterial.Ka = [ li[ 1 ], li[ 2 ], li[ 3 ] ];
    } else if ( li[ 0 ] === 'Kd' ) {
      currentMaterial.Kd = [ li[ 1 ], li[ 2 ], li[ 3 ] ];
    } else if ( li[ 0 ] === 'Ks' ) {
      currentMaterial.Ks = [ li[ 1 ], li[ 2 ], li[ 3 ] ];
    } else if ( li[ 0 ] === 'Ke' ) {
      currentMaterial.Ke = [ li[ 1 ], li[ 2 ], li[ 3 ] ];
    } else if ( li[ 0 ] === 'Ns' ) {
      currentMaterial.Ns = li[ 1 ];
    } else if ( li[ 0 ] === 'd' ) {
      currentMaterial.d = li[ 1 ];
    } else if ( li[ 0 ] === 'map_Kd' ) {
      currentMaterial.map_Kd = li[ 1 ];
    } else if ( li[ 0 ] === 'map_d' ) {
      currentMaterial.map_d = li[ 1 ];
    }
  } );

  // -- check obj index usage ----------------------------------------------------------------------
  const firstIndices = objects[ 0 ].f[ 0 ][ 0 ];
  const indexUsage = [
    firstIndices[ 0 ] != null, // v
    firstIndices[ 1 ] != null, // vt
    firstIndices[ 2 ] != null, // vn
  ];

  // -- obj v/vt to ms3dascii vertex ---------------------------------------------------------------
  vertices.ms = [];
  const indexKeyMap = new Map();

  // attribute vertices with textureCoords
  objects.forEach( ( object ) => {
    object.f.forEach( ( face ) => {
      face.forEach( ( index ) => {
        const indexKey = index.join( '/' );
        if ( indexKeyMap[ indexKey ] == null ) {
          const xyz = vertices.v[ index[ 0 ] - 1 ];
          const uv = indexUsage[ 1 ] ? vertices.vt[ index[ 1 ] - 1 ] : [ 0.0, 0.0 ];

          vertices.ms.push( [
            xyz[ 0 ],
            xyz[ 1 ],
            xyz[ 2 ],
            uv[ 0 ],
            uv[ 1 ],
          ] );

          indexKeyMap[ indexKey ] = vertices.ms.length - 1;
        }
      } );
    } );
  } );

  // -- obj vn to ms3dascii normal -----------------------------------------------------------------
  vertices.msn = [];

  if ( indexUsage[ 2 ] ) {
    vertices.vn.forEach( ( normal ) => {
      vertices.msn.push( [ normal[ 0 ], normal[ 1 ], normal[ 2 ] ] );
    } );
  } else {
    // auto generate normals by face direction
    objects.forEach( ( object, iObject ) => {
      vertices.msn[ iObject ] = [];
      object.f.forEach( ( face ) => {
        const v = face.map( ( i ) => vertices.v[ i[ 0 ] ] );

        const a = [
          v[ 1 ][ 0 ] - v[ 0 ][ 0 ],
          v[ 1 ][ 1 ] - v[ 0 ][ 1 ],
          v[ 1 ][ 2 ] - v[ 0 ][ 2 ],
        ];
        const b = [
          v[ 2 ][ 0 ] - v[ 1 ][ 0 ],
          v[ 2 ][ 1 ] - v[ 1 ][ 1 ],
          v[ 2 ][ 2 ] - v[ 1 ][ 2 ],
        ];
        const n = [
          a[ 1 ] * b[ 2 ] - a[ 2 ] * b[ 1 ],
          a[ 2 ] * b[ 0 ] - a[ 0 ] * b[ 2 ],
          a[ 0 ] * b[ 1 ] - a[ 1 ] * b[ 0 ],
        ];
        const ln = -Math.sqrt( n[ 0 ] * n[ 0 ] + n[ 1 ] * n[ 1 ] + n[ 2 ] * n[ 2 ] ); // wtf is this minus
        const nn = n.map( ( v ) => v / ln );

        vertices.msn[ iObject ].push( nn );
      } );
    } );
  }

  // -- write ms3dascii ----------------------------------------------------------------------------
  let out = `// MilkShape 3D ASCII
// Converted from .obj with obj2ms3dascii.lua by 0b5vr
// https://github.com/0b5vr/obj2ms3dascii
Frames: 30
Frame: 1
`;

  out += `Meshes: ${ objects.length }\n`;

  objects.forEach( ( object, iObject ) => {
    const materialIndex = materials.findIndex( ( material ) => material.name === object.usemtl );
    out += `"${ object.name }" 0 ${ materialIndex }\n`;

    out += `${ vertices.ms.length }\n`;
    vertices.ms.forEach( ( vertex ) => {
      out += `0 ${ vertex.join( ' ' ) } -1\n`;
    } );

    if ( indexUsage[ 2 ] ) {
      out += `${ vertices.msn.length }\n`;
      vertices.msn.forEach( ( normal ) => {
        out += `${ normal.join( ' ' ) }\n`;
      } );
    } else {
      out += `${ vertices.msn[ iObject ].length }\n`;
      vertices.msn[ iObject ].forEach( ( normal ) => {
        out += `${ normal.join( ' ' ) }\n`;
      } );
    }

    out += `${ object.f.length }\n`;
    if ( indexUsage[ 2 ] ) {
      object.f.forEach( ( face ) => {
        out += `0 ${ face.map( ( index ) => indexKeyMap[ index.join( '/' ) ] ).join( ' ' ) } ${ face.map( ( index ) => index[ 2 ] - 1 ).join( ' ' ) } 1\n`;
      } );
    } else {
      object.f.forEach( ( face, iFace ) => {
        out += `0 ${ face.map( ( index ) => indexKeyMap[ index.join( '/' ) ] ).join( ' ' ) } ${ iFace } ${ iFace } ${ iFace } 1\n`;
      } );
    }
  } );

  out += `Materials: ${ materials.length }\n`;
  materials.forEach( ( material ) => {
    out += `"${ material.name }"
${ material.Ka.join( ' ' ) } 1
${ material.Kd.join( ' ' ) } 1
${ material.Ks.join( ' ' ) } 1
${ material.Ke.join( ' ' ) } 1
${ material.Ns }
${ material.d }
"${ material.map_Kd }"
"${ material.map_d }"
`;
  } );

  out += `Bones: 1
"Bone01"
""
0 0 0 0 0 0 0
1
1 0 0 0
1
1 0 0 0

GroupComments: 0
MaterialComments: 0
BoneComments: 0
ModelComment: 0`;

  return out;
}
